demos/common/hfsm/az_iot_hfsm.c

```diff
@@ -0,0 +1,324 @@
+          ret = az_hfsm_post_event(this_iothfsm->_provisioning_hfsm, az_hfsm_event_az_iot_start);
+          ret = az_hfsm_transition_substate(me, azure_iot, provisioning);
+#else
+          ret = az_hfsm_post_event(this_iothfsm->_iothub_hfsm, az_hfsm_event_az_iot_start);
+          ret = az_hfsm_transition_substate(me, azure_iot, hub);
+#endif
+      }
+      break;
+
+    case AZ_HFSM_EXIT:
+    case AZ_HFSM_ERROR:
+    case AZ_HFSM_TIMEOUT:
+    default:
+      LogInfo( ("AzureIoT: PANIC!") );
+      az_iot_hfsm_pal_critical_error(me);
+      break;
+  }
+
+  return ret;
+}
+
+static int32_t idle(az_hfsm* me, az_hfsm_event event)
+{
+  int32_t ret = 0;
+  az_iot_hfsm_type* this_iothfsm = (az_iot_hfsm_type*)me;
+
+  switch ((int32_t)event.type)
+  {
+    case AZ_HFSM_ENTRY:
+      LogInfo( ("idle: Entry") );
+      break;
+
+    case AZ_HFSM_EXIT:
+      LogInfo( ("idle: Exit") );
+      break;
+    
+    case AZ_IOT_START:
+      LogInfo( ("idle: AZ_IOT_START") );
+
+#ifdef AZ_IOT_HFSM_PROVISIONING_ENABLED
+      if(az_hfsm_post_event(this_iothfsm->_provisioning_hfsm, az_hfsm_event_az_iot_start))
+      {
+        LogError( ("idle: az_hfsm_post_event to _provisioning_hfsm failed.") );
+        ret = az_hfsm_post_event(me, az_hfsm_errork_unknown_event);
+      }
+      else
+      {
+        ret = az_hfsm_transition_peer(me, idle, provisioning);
+      }
+#else
+      if(az_hfsm_post_event(this_iothfsm->_iothub_hfsm, az_hfsm_event_az_iot_start))
+      {
+        LogError( ("idle: az_hfsm_post_event to _iothub_hfsm failed.") );
+        ret = az_hfsm_post_event(me, az_hfsm_errork_unknown_event);
+      }
+      else
+      {
+        ret = az_hfsm_transition_peer(me, idle, hub);
+      }
+#endif
+      break;
+
+    default:
+      ret = AZ_HFSM_RET_HANDLE_BY_SUPERSTATE;
+  }
+
+  return ret;
+}
+
+#ifdef AZ_IOT_HFSM_PROVISIONING_ENABLED
+// AzureIoT/Provisioning
+static int32_t provisioning(az_hfsm* me, az_hfsm_event event)
+{
+  int32_t ret = 0;
+  az_iot_hfsm_type* this_iothfsm = (az_iot_hfsm_type*)me;
+
+  switch ((int32_t)event.type)
+  {
+    case AZ_HFSM_ENTRY:
+      LogInfo( ("provisioning: Entry") );
+      this_iothfsm->_retry_attempt = 0;
+      this_iothfsm->_start_time_msec = az_hfsm_pal_timer_get_milliseconds();
+      this_iothfsm->_timer_handle = az_hfsm_pal_timer_create(me);
+      break;
+
+    case AZ_HFSM_EXIT:
+      LogInfo( ("provisioning: Exit") );
+      az_hfsm_pal_timer_destroy(me, this_iothfsm->_timer_handle);
+      break;
+
+    case AZ_HFSM_TIMEOUT:
+      LogInfo( ("provisioning: Timeout") );
+      this_iothfsm->_start_time_msec = az_hfsm_pal_timer_get_milliseconds();
+      ret = az_hfsm_post_event(this_iothfsm->_provisioning_hfsm, az_hfsm_event_az_iot_start);
+      break;
+    
+    case AZ_IOT_PROVISIONING_DONE:
+      LogInfo( ("provisioning: Done") );
+      ret = az_hfsm_post_event(this_iothfsm->_iothub_hfsm, az_hfsm_event_az_iot_start);
+      ret = az_hfsm_transition_peer(me, provisioning, hub);
```

Comment:
previous ret is overridden

Reply:
will remove transition returns. The post may fail so it will keep the "ret" (queue full if queues are used for example).